src/gluonts/shell/serve/__init__.py

```diff
@@ -0,0 +1,169 @@
+from flask import Flask
+from gunicorn.app.base import BaseApplication
+from pydantic import BaseSettings
+
+# First-party imports
+import gluonts
+from gluonts.core import fqname_for
+from gluonts.core.component import check_gpu_support
+from gluonts.model.estimator import Estimator
+from gluonts.model.predictor import Predictor
+from gluonts.shell.sagemaker import ServeEnv
+
+from .app import make_app
+
+logging.basicConfig(
+    level=logging.INFO,
+    format="%(asctime)s [%(process)d] [%(levelname)s] %(message)s",
+    datefmt="[%Y-%m-%d %H:%M:%S %z]",
+)
+
+logger = logging.getLogger("gluonts.serve")
+
+MB = 1024 * 1024
+
+
+class Settings(BaseSettings):
+    # see: https://pydantic-docs.helpmanual.io/#settings
+    class Config:
+        env_prefix = ""
+
+    model_server_workers: Optional[int] = None
+    max_content_length: int = 6 * MB
+
+    sagemaker_server_address: IPv4Address = IPv4Address("0.0.0.0")
+    sagemaker_server_port: int = 8080
+    sagemaker_server_timeout: int = 100
+
+    sagemaker_batch: bool = False
+    sagemaker_batch_strategy: str = "SINGLE_RECORD"
+
+    sagemaker_max_payload_in_mb: int = 6
+    sagemaker_max_concurrent_transforms: int = 2 ** 32 - 1
+
+    @property
+    def sagemaker_server_bind(self) -> str:
+        return f"{self.sagemaker_server_address}:{self.sagemaker_server_port}"
+
+    @property
+    def number_of_workers(self) -> int:
+        cpu_count = multiprocessing.cpu_count()
+
+        if self.model_server_workers:
+            logging.info(
+                f"Using {self.model_server_workers} workers "
+                f"(set by MODEL_SERVER_WORKERS environment variable)."
+            )
+            return self.model_server_workers
+
+        elif (
+            self.sagemaker_batch
+            and self.sagemaker_max_concurrent_transforms < cpu_count
+        ):
+            logger.info(
+                f"Using {self.sagemaker_max_concurrent_transforms} workers "
+                f"(set by MaxConcurrentTransforms parameter in batch mode)."
+            )
+            return self.sagemaker_max_concurrent_transforms
+
+        else:
+            logger.info(f"Using {cpu_count} workers")
+            return cpu_count
+
+
+class Application(BaseApplication):
+    def __init__(self, app, config) -> None:
+        self.app = app
+        self.config = config
+        BaseApplication.__init__(self)
+
+    def load_config(self) -> None:
+        for key, value in self.config.items():
+            if key in self.cfg.settings and value is not None:
+                self.cfg.set(key, value)
+
+    def init(self, parser, opts, args):
+        pass
+
+    def load(self) -> Flask:
+        return self.app
+
+    def stop(self, *args, **kwargs):
+        logger.info("Shutting down GluonTS scoring service")
+
+
+def make_gunicorn_app(
+    env: ServeEnv,
+    forecaster_type: Optional[Type[Union[Estimator, Predictor]]],
+    settings: Settings,
+) -> Application:
+    check_gpu_support()
```

Comment:
I guess for supporting gpu's properly, we need a bit more logic for handling the number of proceses.

Also, the server should probably be started with the following environment variables to ensure that each server has one cpu and mxnet does not try to parallelize a single request. 
```
OMP_NUM_THREADS=1
MXNET_ENGINE_TYPE=NaiveEngine
```

Reply:
Thanks. These changes were just moved around, but we should still do what you’ve suggested.